Adding CDM and DSBulk Migrator details (#144)
datastax/migration-docs: docs-src/zdm-core/modules/migrate/pages/cassandra-data-migrator.adoc

Use {cstar-data-migrator} to migrate and validate tables between Origin and Target Cassandra clusters, with available logging and reconciliation support.

[[cdm-prereqs]]
== {cstar-data-migrator} prerequisites

* Install or switch to Java 8. The Spark binaries are compiled with this version of Java.
* Install https://archive.apache.org/dist/spark/spark-3.3.1/[Spark 3.3.1^] on a single VM (no cluster necessary) where you want to run this job. 
* Optionally, install https://maven.apache.org/download.cgi[Maven^] 3.8.x, if you want to build the JAR for local development.

You can install Apache Spark by running the following commands:

[source,bash]
----
wget https://archive.apache.org/dist/spark/spark-3.3.1/spark-3.3.1-bin-hadoop3.tgz 

tar -xvzf spark-3.3.1-bin-hadoop3.tgz
----

[[cdm-install-as-container]]
== Install {cstar-data-migrator} as a Container

Get the latest image that includes all dependencies from https://hub.docker.com/r/datastax/cassandra-data-migrator[DockerHub^].

All migration tools (`cassandra-data-migrator` + `dsbulk` + `cqlsh`) are available in the `/assets/` folder of the container.

[[cdm-install-as-jar]]
== Install {cstar-data-migrator} as a JAR file

> Download the *latest* JAR file from the {cstar-data-migrator} https://github.com/datastax/cassandra-data-migrator/packages/[GitHub repo^]. 

Reviewer: ```suggestion
Download the *latest* JAR file from the {cstar-data-migrator} https://github.com/datastax/cassandra-data-migrator/packages/[GitHub repo^]. image:https://img.shields.io/github/v/release/datastax/cassandra-data-migrator?color=green[Latest release]
```

[NOTE]
====

> Version 4.x of {cstar-data-migrator} is not backward-compatible with *.properties files created in previous versions, and package names have changed. As of 21-Jun-2023, 4.1.0 is the latest version.

Reviewer: ```suggestion
Version 4.x of {cstar-data-migrator} is not backward-compatible with `*.properties` files created in previous versions, and package names have changed. As of 21-Jun-2023, `4.1.0` is the latest version. If you're starting new, it is recommended to use the latest release version.
```

====

[[cdm-build-jar-local]]
== Build {cstar-data-migrator} JAR for local development (optional)

Optionally, you can build the {cstar-data-migrator} JAR for local development. (You'll need https://maven.apache.org/download.cgi[Maven^] 3.8.x.)

Example:

[source,bash]
----
cd ~/github
git clone git@github.com:datastax/cassandra-data-migrator.git
cd cassandra-data-migrator
mvn clean package
----

> The fat jar (`cassandra-data-migrator-4.x.x.jar`) file should be present now in the `target` folder.

Reviewer: Let's future-proof it 👍🏼 
```suggestion
The fat jar (`cassandra-data-migrator-x.y.z.jar`) file should be present now in the `target` folder.
```

   * The simplified sample properties configuration, https://github.com/datastax/cassandra-data-migrator/blob/main/src/resources/cdm.properties[cdm.properties^]. This file contains only those parameters that are commonly configured.
   * The complete sample properties configuration, https://github.com/datastax/cassandra-data-migrator/blob/main/src/resources/cdm-detailed.properties[cdm-detailed.properties^], for the full set of configurable settings.

2. Place the properties file that you elected to use and customize where it can be accessed while running the job via `spark-submit`.

3. Run the job using `spark-submit` command:

[source,bash]
----
./spark-submit --properties-file cdm.properties /
--conf spark.cdm.schema.origin.keyspaceTable="<keyspacename>.<tablename>" /
--master "local[*]" /

> --class com.datastax.cdm.job.Migrate cassandra-data-migrator-4.x.x.jar &> logfile_name_$(date +%Y%m%d_%H_%M).txt

Reviewer: ```suggestion
--class com.datastax.cdm.job.Migrate cassandra-data-migrator-x.y.z.jar &> logfile_name_$(date +%Y%m%d_%H_%M).txt
```

----

[TIP]
====
* The `spark-submit` command generates a log file, `logfile_name_*.txt`, to avoid log output on the Terminal console.
* If the table you're migrating is large (such as over 100GB), you can add the option `--driver-memory 25G --executor-memory 25G`. Example:

[source,bash]
----
./spark-submit --properties-file cdm.properties /
--conf spark.cdm.schema.origin.keyspaceTable="<keyspacename>.<tablename>" /
--master "local[*]" --driver-memory 25G --executor-memory 25G /

> --class com.datastax.cdm.job.Migrate cassandra-data-migrator-4.x.x.jar &> logfile_name_$(date +%Y%m%d_%H_%M).txt

Reviewer: ```suggestion
--class com.datastax.cdm.job.Migrate cassandra-data-migrator-x.y.z.jar &> logfile_name_$(date +%Y%m%d_%H_%M).txt
```

----
====

[[cdm-validation-steps]]
== {cstar-data-migrator} steps in validation mode

To run your migration job with {cstar-data-migrator} in **data validation mode**, use class option `--class com.datastax.cdm.job.DiffData`. 
Example:

[source,bash]
----
./spark-submit --properties-file cdm.properties /
--conf spark.cdm.schema.origin.keyspaceTable="<keyspacename>.<tablename>" /
--master "local[*]" /

> --class com.datastax.cdm.job.DiffData cassandra-data-migrator-4.x.x.jar &> logfile_name_$(date +%Y%m%d_%H_%M).txt

Reviewer: ```suggestion
--class com.datastax.cdm.job.DiffData cassandra-data-migrator-x.y.z.jar &> logfile_name_$(date +%Y%m%d_%H_%M).txt
```

spark.cdm.autocorrect.mismatch                    false|true
----

[IMPORTANT]
====
The {cstar-data-migrator} validation job will never delete records from Target. The job only adds or updates data on Target.
====

[[cdm--partition-ranges]]
== Migrating or validating specific partition ranges

You can also use {cstar-data-migrator} to migrate or validate specific partition ranges, by using a **partition-file** with the name `./<keyspacename>.<tablename>_partitions.csv`. Use the following format in the CSV file, in the current folder as input. 
Example:

[source,csv]
----
-507900353496146534,-107285462027022883
-506781526266485690,1506166634797362039
2637884402540451982,4638499294009575633
798869613692279889,8699484505161403540
----

Each line in the CSV represents a partition-range (`min,max`). 

Alternatively, you can also pass the partition-file via a command-line parameter. 
Example:

[source,bash]
----
./spark-submit --properties-file cdm.properties /
 --conf spark.cdm.schema.origin.keyspaceTable="<keyspacename>.<tablename>" /
 --conf spark.cdm.tokenRange.partitionFile="/<path-to-file>/<csv-input-filename>" /
 --master "local[*]" /

>  --class com.datastax.cdm.job.<Migrate|DiffData> cassandra-data-migrator-4.x.x.jar &> logfile_name_$(date +%Y%m%d_%H_%M).txt

Reviewer: ```suggestion
 --class com.datastax.cdm.job.<Migrate|DiffData> cassandra-data-migrator-x.y.z.jar &> logfile_name_$(date +%Y%m%d_%H_%M).txt
```
